### PR TITLE
[StubIr][Interop] Structs support

### PR DESCRIPTION
**Pull request by sbogolepov**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGen.kt [hunk 1]
```diff
@@ -610,6 +610,8 @@ private fun IrType.isCEnumType(): Boolean {
 // Make sure external stubs always get proper annotaions.
 private fun IrDeclaration.hasCCallAnnotation(name: String): Boolean =
         this.annotations.hasAnnotation(cCall.child(Name.identifier(name)))
+                // LazyIr doesn't pass annotations from descriptor to IrValueParameter.
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGen.kt]: It makes sense to get this eventually fixed.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/cstruct/CStructVarClassGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/cstruct/CStructVarClassGenerator.kt]: Isn't 2020 already?
Same below.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 3]
```diff
@@ -38,10 +36,20 @@ private fun isEnumVarValueAccessor(function: IrFunction, symbols: KonanSymbols):
...
+
+private fun isBitFieldAccessor(function: IrFunction): Boolean =
+        function.hasAnnotation(RuntimeNames.cStructBitField)
+
 private class InteropCallContext(
         val symbols: KonanSymbols,
         val builder: IrBuilderWithScope
-)
+) {
+    fun IrType.isCPointer(): Boolean = isSubtypeOfClass(symbols.interopCPointer)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Is subtyping check really required here?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 4]
```diff
@@ -154,40 +162,165 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
...
+            .functions.single { it.name.identifier == "plus" }
+    return with (builder) {
+        irCall(nativePtrPlusLong).also {
+            it.dispatchReceiver = base
+            it.putValueArgument(0, irLong(offset))
+        }
+    }
+}
+
+private fun InteropCallContext.readPointerFromMemory(nativePtr: IrExpression): IrExpression {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Doesn't seem to read anything from memory.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 5]
```diff
@@ -154,40 +162,165 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
...
         }
         accessor.isSetter -> {
-            val arg = callSite.getValueArgument(0)!!
-            val valueToWrite = convertEnumToIntegral(arg)
-            writePrimitiveToMemory(nativePtr, valueToWrite)
+            val value = callSite.getValueArgument(0)!!
+            when {
+                value.type.isCEnumType() -> writeEnumValueToMemory(fieldPointer, value)
+                value.type.isPrimitiveType() -> writePrimitiveToMemory(fieldPointer, value)
+                else -> error("Cannot set field of type ${value.type.getClass()?.name}")
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Pointer-typed struct fields?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 6]
```diff
@@ -154,40 +162,165 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
...
+                else -> error("Cannot get field type: ${type.getClass()?.name}")
+            }
         }
         accessor.isSetter -> {
-            val arg = callSite.getValueArgument(0)!!
-            val valueToWrite = convertEnumToIntegral(arg)
-            writePrimitiveToMemory(nativePtr, valueToWrite)
+            val value = callSite.getValueArgument(0)!!
+            when {
+                value.type.isCEnumType() -> writeEnumValueToMemory(fieldPointer, value)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Using property type or setter's parameter type is more reliable.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 7]
```diff
@@ -154,40 +162,165 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
...
+    }
+}
+
+private fun InteropCallContext.readBits(
+        base: IrExpression,
+        offset: Long,
+        size: Int,
+        type: IrType
+): IrExpression {
+    val isSigned = !type.isUnsigned()
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: What if `type` is enum?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 4 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: 🤦‍♂️

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 7 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Btw, do we have a test that would discover that bug?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt [hunk 8]
```diff
@@ -154,40 +162,192 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
...
-            val enumClass = accessor.returnType.getClass()!!
-            val enumPrimitiveType = enumClass.properties.single { it.name.asString() == "value" }
-                    .getter!!.returnType
-            val readMemory = readPrimitiveFromMemory(nativePtr, enumPrimitiveType)
-            return convertIntegralToEnum(readMemory, accessor.returnType)
+            val type = accessor.returnType
+            when {
+                type.isCEnumType() -> readEnumValueFromMemory(fieldPointer, type)
+                type.isPrimitiveType() -> readPrimitiveFromMemory(fieldPointer, type)
+                type.isCPointer() -> readPointerFromMemory(fieldPointer)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Does this work for array-typed members of structs?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 7 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: `interop_bitfields` covers this case, so yes. Looks like I forgot to turn `-mode metadata` on for this particular test.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 8 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: It should, but it won't work without #3826 that adds support for `CArrayPointer` typealias.

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Getting array-typed member doesn't involve reading from memory, however this branch does perform reading. How can this work?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 7 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: It seems to contain only one enum. We probably should have both signed and unsigned enum bitfield tests.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt (hunk 8 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt]: Ok, I see.

File: backend.native/tests/interop/basics/bitfields.def [hunk 9]
```diff
@@ -3,13 +3,18 @@ enum B2 {
...
+};
+
 struct __attribute__((packed)) S {
     long long x1 : 1;
     enum B2 x2 : 2;
     unsigned short x3 : 3;
     unsigned int x4 : 4;
     int x5 : 5;
     long long x6 : 63;
+    enum E x7: 1;
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/interop/basics/bitfields.def]: Doesn't make much sense without testing it from Kotlin code. See `bf.kt`.

**Comment by SvyatoslavScherbina** [comment]
We definitely need test coverage for different field types, maybe later.

File: backend.native/tests/interop/basics/bitfields.def (hunk 9 above)
Comment (sbogolepov) [reply backend.native/tests/interop/basics/bitfields.def]: Yeah, for reason unknown `bf.kt` changes didn't make it to GitHub ¯\_(ツ)_/¯.

**Pull request merged**